### DISCUSSION TITLE
How to prevent type widening?

### DISCUSSION THREAD

**Discussion by patrick-kidger** [microsoft/pyright#5597] [Q&A]
Considering this example from the docs (https://microsoft.github.io/pyright/#/mypy-comparison?id=constraint-solver-type-widening):
```python
T = TypeVar("T")
def func(val1: T, val2: T) -> T:
    ...

func("", 1)
```
what are the appropriate incantations to have pyright treat this as an error? In my use-case, `T` can be any (concrete) type at all, but it must be the same (concrete) type for both arguments.

I'm aware of `TypeVar(... *constraints)` to get this behaviour, but this only works if you have some finite set of allowed types, that is known ahead of time.

**Comment by erictraut** [comment]
You would need to use a constrained type variable if you want to constrain `T` to a specific set of types. Otherwise the constraint solver is free to use type widening (pyright) or type joins (mypy) in its pursuit to find a type that satisfies all of the argument types. In your example, pyright successfully solves `T` as `str | int`. Mypy also successfully finds a solution: `object`. There's no way when using a non-constrained TypeVar to tell the constraint solver not to widen types in its pursuit of a solution. As long as _some_ solution can be found, no error will be reported.

For details, refer to [this section of PEP 483](https://peps.python.org/pep-0483/#type-variables).

**Reply by patrick-kidger** [reply]
Got it. That's unfortunate that there's no way to convey this kind of constraint information to the type checker. Thank you!
